dashboard/app/models/artist.rb

```diff
@@ -264,4 +275,44 @@ def k1_blocks
       <block type="simple_jump_left"></block>
     XML
   end
+
+  def strip_block(block)
+    stripped_block = block.dup
+    stripped_block.xpath(STRIPPED_NODES_XPATH).remove
+    stripped_block['type'] = stripped_block['type'].chomp '_dropdown'
+    STRIPPED_ATTRS.each {|attr| stripped_block.remove_attribute(attr)}
+    stripped_block.content = stripped_block.content.strip
+    return stripped_block.to_xml
+  end
+
+  # Add blocks to the toolbox that appear in the solution, but aren't already
+  # in the toolbox
+  def add_missing_toolbox_blocks
+    toolbox = Nokogiri::XML(properties['toolbox_blocks'])
+    toolbox_blocks = toolbox.xpath('//block')
+    Nokogiri::XML(properties['solution_blocks']).xpath('//block').each do |block|
+      next if IGNORED_SOLUTION_BLOCK_ATTRS.any? {|kvpair| block.attr(kvpair[0]) == kvpair[1]}
+
+      stripped_block = strip_block block
+      next if toolbox_blocks.any? do |toolbox_block|
+        stripped_block == strip_block(toolbox_block)
+      end
+
+      # Solution block does not appear in the toolbox, add it
+      toolboxified_block = block.dup
+      toolboxified_block.xpath(STRIPPED_NODES_XPATH).remove
+      toolboxified_block.content = toolboxified_block.content.strip
```

Comment:
any way this could be combined with strip_block? They're doing some similar things and some different things, but it sure feels like there's a way to avoid repeating ourselves.